docs/cache-server.md

```diff
@@ -0,0 +1,125 @@
+### Purpose
+
+The primary purpose of the cache server is to support cross-shard communication.
+Shards need to communicate with each other in order to enable, among other things, migration and replication features.
+Direct shard-to-shard communication is not feasible in a larger setup.
+The web of connections would be hard to reason about, maintain and troubleshoot. 
+Thus, the cache server serves as a central place for shards to store and read common data.
+
+### High-level overview
+
+The cache server is a regular, Kubernetes-style CRUD API server with support of LIST/WATCH semantics. 
+Conceptually it supports two modes of operations. 
+The first mode is in which a shard gets its private space for storing its own data. 
+The second mode is in which a shard can read other shards' data.
+
+The first mode of operation is implemented as a controller that runs within a shard. 
+It maintains an informer for `LIST/WATCH`ing its own resources. 
+It compares received data to its local copy and makes sure both copies are consistent.
+
+The second mode of operation is implemented as a separate informer within a shard. 
+Thanks to that a shard can implement a different resiliency strategy. 
+For example, it can tolerate the unavailability of the secondary during startup and become ready.
+
+### Running the server
+
+The cache server can be run as a standalone binary or as part of a kcp server.
+
+The standalone binary is in https://github.com/kcp-dev/kcp/tree/main/cmd/cache-server and can be run by issuing `go run ./cmd/cache-server/main.go` command.
+
+To run it as part of a kcp server, pass `--cache-url` flag to the kcp binary.
+
+### Client-side functionality
+
+In order to interact with the cache server from a shard, the https://github.com/kcp-dev/kcp/tree/main/pkg/cache/client 
+repository provides the following client-side functionality:
+
+[ShardRoundTripper](https://github.com/kcp-dev/kcp/blob/b739fa5b5c83fb2c43b631c6234264d5dd1fc6e4/pkg/cache/client/round_tripper.go#L54), 
+a shard aware wrapper around `http.RoundTripper`. It changes the URL path to target a shard from the context.
+
+[DefaultShardRoundTripper](https://github.com/kcp-dev/kcp/blob/b739fa5b5c83fb2c43b631c6234264d5dd1fc6e4/pkg/cache/client/round_tripper.go#L128) 
+is a `http.RoundTripper` that sets a default shard name if not specified in the context.
+
+For example, in order to make a client shard aware, inject the `http.RoundTrippers` to a `rest.Config`
+
+```
+import(
+cacheclient "github.com/kcp-dev/kcp/pkg/cache/client"
```

Comment:
run `go fmt` / provide an example snippet here that someone could copy-paste please

Reply:
I don't have a complete example. I will add one, once we will write a controller consuming data from the cache.